bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/IStructuredSelection.java

```diff
@@ -15,27 +15,92 @@
 
 import java.util.Iterator;
 import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
 
 /**
  * A selection containing elements.
  */
 @SuppressWarnings("rawtypes")
 public interface IStructuredSelection extends ISelection, Iterable {
```

Comment:
I have the impression that with all the default implementations we now have an endless loop (`iterator()` uses `stream()` and `stream()` the spliterator(), which uses `iterator()`). I think it would be good to enforce at least one implementation.
`iterator()` is probably always possible in a simple way.

Reply:
stream should be the method to actually implement, everything else could be derived from it... but I would leave out that for now anyways and split this into smaller chunks.